[LSROA] Add logical SROA pass (#192058)
llvm/llvm-project: llvm/docs/LangRef.rst

The resulting pointer belongs to the same address space as ``source``.
This instruction does not dereference the pointer.

> Chaining indices in a single ``llvm.structured.gep`` instruction or spliting

Reviewer: ```suggestion
Chaining indices in a single ``llvm.structured.gep`` instruction or splitting
```

> is sequentialy over multiple has exactly the same behavior. As such, two

Reviewer: ```suggestion
it sequentially over multiple instructions has exactly the same behavior. As such, two
```

> ``llvm.structured.gep`` instruction can be merged if the second one base

Reviewer: ```suggestion
``llvm.structured.gep`` instructions can be merged if the second one base
```

pointer is the result of the first by appending the indices of the second to
the first.

> Conversely, a single structured GEP instruction can be split in to by

Reviewer: ```suggestion
Conversely, a single structured GEP instruction can be split in two by
```

> splitting the indices sequence over multiple instructions.

Reviewer: In general, I don't like the way this phrased. Whether an optimization is legal or not should not be done by fiat. It should be corollary of the definition. I believe that in this case it is. If other disagree we should try to clean up the definition above. 

You define the way the intrinsic is that it computes pointers iteratively. Each iteration produces a pointer to a memory location that hold an object of a particular type or poison if out of bounds.

This paragraph should explain that it is allowed as long at the types match properly. Reference the earlier text to explain why that is true. Give an example of splitting that is not allowed, explaining how it changes the semantics. Like changing the size of the array.

Author: To me this paragraph is a no-op definition-wise, but make something that is inferred explicit:

In the current definition:
```
The first index determines which element/field of ``basetype`` is selected,
computes the pointer to access this element/field assuming ``source`` points
to the start of ``basetype``.
This pointer becomes the new ``source``, the current type the new
``basetype``, and the next indices is consumed until a scalar type is
reached or all indices are consumed.
```

IMO this already defines the iterative nature of the indices, and already validates a structured GEP can be split/merged.
This paragraph is more or less explicitly telling the same thing.
@nikic what was the thing you were worried about?

Reviewer 2: My concern is not related to splitting the GEP, but to how the resulting pointer is used.

I believe that nothing in the current definition of structural.gep prevents you from writing something like this:
```
%p.0 = structural.gep {i32, i32}, %ptr, 0
%p.1 = ptradd %p.0, 4
load %p.1
```

If this code is legal, then splitting the alloca into two is invalid, as the ptradd crosses to the second element.

What I said about subobject provenance is making that `load` illegal, because you only have provenance for the i32 element.

Author: Ahh right, mixing usages. 
Overall, I want to prevent those mixed logical/physical usages.
I'll see how I can formulate this

Reviewer 2: I mean, it doesn't even have to be a mixed logical/physical usage, it could also just be two structural.gep with mismatching types.

Author: Gave this a bit of a thought, and I don't think we should say it's illegal: I think this optimization should instead validate all pointer users (and indirect) are fine and that the pointer is not escaping.

This means we should only split alloca if we see the whole load/store chain, and that we don't pass the pointer to a call.
I'll add tests for such cases, and modify the pass.

Reviewer 2: Yes, that would match what normal SROA does. But I think I saw some HLSL use case where they need SROA to work even if the pointer escapes... not sure where tho.

Author: I think in most cases, at least SPIR-V, we should be fine by running LSROA after inlining. For module linking, I'll need to check, but worse case we under-optimize and will have to handle more cases/refine langref later on.

Reviewer 2: I found what I had in mind: https://github.com/llvm/llvm-project/pull/161601 cc @Icohedron

Looking at the IR in https://github.com/llvm/llvm-project/issues/147109 it *might* still be fine because that doesn't have an actual escape, just a variable index, and that should still be supportable.

Author: Ah right, yes this is fine, at least we extract the array from the parent struct. I added tests for this case too.
For now, this pass doesn't split arrays, but that's something we can add (#192620)

The resulting pointer belongs to the same address space as ``source``.
This instruction does not dereference the pointer.

Aliasing rules:
"""""""""""""""

Common :ref:`aliasing rules <pointeraliasing>` apply to pointers returned
by this intrinsic, as well as the following additional rules:

- A pointer returned by `@llvm.structured.gep` for a given element is
  considered not to alias with pointers the instruction would return for

>   a sibling element.

Reviewer: ```suggestion
A pointer returned by `@llvm.structured.gep` is considered not to alias a pointer returned by another `@llvm.structured.gep` with the same base pointer unless the index sequence of one is a prefix of the other.
```

Author: Ah I like the "prefix" for indices. Better than siblings

.. code-block:: llvm

   %S = { i32, i32 }
   %ptr = call ptr @llvm.structured.gep(ptr elementtype(%S) %src, i32 0)

>    %res = load i64, ptr %ptr

Reviewer: What are you trying to say here? Aliasing is a statement about two pointers. Two pointers alias if the point to the same memory. In this example, you only have a single pointer. 

For LSROA, I think we want to be able to say that `%ptr0` and `%ptr1` below do not alias. Then the compiler is allowed to rearrange the memory.

```
   %ptr0 = call ptr @llvm.structured.gep(ptr elementtype(%S) %src, i32 0)
   %ptr1 = call ptr @llvm.structured.gep(ptr elementtype(%S) %src, i32 1)
```

Author: The second pointer is implied:
What we this explains is the you cannot load the second field of `%S` through `ptr`, even by loading 64 bits (which would be possible for a normal GEP)
Rephrased it to make the second pointer explicit.

The resulting pointer belongs to the same address space as ``source``.
This instruction does not dereference the pointer.

Aliasing rules:
"""""""""""""""

Common :ref:`aliasing rules <pointeraliasing>` apply to pointers returned
by this intrinsic, as well as the following additional rules:

A pointer returned by `@llvm.structured.gep` is considered not to alias a
pointer returned by another `@llvm.structured.gep` with the same base pointer
unless the index sequence of one is a prefix of the other.

.. code-block:: llvm

   %S = { i32, i32 }
   %ptr0 = call ptr @llvm.structured.gep(ptr elementtype(%S) %src, i32 0)
   %ptr1 = call ptr @llvm.structured.gep(ptr elementtype(%S) %src, i32 1)

   %field0 = load i64, ptr %ptr0
   %field1 = load i32, ptr %ptr1

Here, `%field0` is not expected to load the second `i32` of the struct `%S`.
The loaded value is implementation defined, and optimizations are free

> to assume there is no overlap in the memory loaded by the 2 loads.

Reviewer 2: This needs to be UB, not implementation defined.

We should also avoid defining things in terms of whether two pointers "alias" -- aliasing is a property that is derived from certain combinations of accesses being UB, not something that's directly part of the semantics.

Here would be my stab at defining this:

> The pointer returned by `@llvm.structured.gep` can only be used to access memory that is part of the indexed subobject, otherwise the behavior is undefined. For example:
>
>     %S = { i32, i32 } ; assume these are laid out next to each other
>     %ptr0 = call ptr @llvm.structured.gep(ptr elementtype(%S) %src, i32 0)
>     %field0 = load i64, ptr %ptr0 ; undefined behavior, because the access crosses
>                                   ; into the second struct element
>
> This implies that two  `@llvm.structured.gep` calls with the same base pointer and element type do not alias unless the index sequence of one is a prefix of the other.

Author: I like this formulation, changed, thanks!
